design-proposals/amd-sev.md

```diff
@@ -0,0 +1,253 @@
+
+The construction of the launch blob is currently not in the scope of KubeVirt as it requires the guest owner to expose the secrets. It should be performed 'offline' by the guest owner (e.g. by using `sev-tool --generate_launch_blob ...`).
+
+### Step #3 Retrieving the measurement of a running VM
+
+The measurement of the memory is performed before a VM is booted. Initially a VM is launched in a paused state thus allowing to retrieve and validate the measurement. KubeVirt can provide the `/sevquerylaunchmeasure` VMI endpoint complemented by `virtctl sevquerylaunchmeasure ...` command to expose it to the guest owner.
+
+The implementation will need to call the corresponding libvirt API to retrieve the measurement for a particular VM:
+
+- [virDomainGetLaunchSecurityInfo](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainGetLaunchSecurityInfo)
+
+    The call returns the launch measurement for a SEV guest.
+
+Currently the measurement is calculated basing on the negotiated session parameters and SHA256 digest of the OVMF binary. The guest owner can verify it against the binary provided by the platform owner.
+
+### Step #4 Launching the VM by injecting the secret
+
+After the measurement is successfully verified, the guest owner can inject the encrypted secret into the VM and resume its execution. A new `/sevinjectsecret` VMI endpoint and `virtctl sevinjectsecret ...` command can be added to support secret injection in KubeVirt VMs.
+
+The implementation will need to call the corresponding libvirt API once it becomes available:
+
+- [sev-inject-launch-secret](https://listman.redhat.com/archives/libvir-list/2021-May/msg00196.html)
+
+    The libvirt API is still missing.
+
+The blob with the encrypted secret can be built by the guest owner using `sev-tool --package_secret ...`.
+
+## Known limitations
+
+- SEV is only going to work with OVMF (UEFI)
+- SEV-encrypted VMs cannot contain directly-accessible host devices (that is, PCI passthrough)
+- Live Migration is not supported with SEV at the moment
+- The ammount of running SEV VMs on a node is limited
+- libvirt currently does not provide an API to inject a secret in a running VM
+
+Some of the limitations will probably be removed in the future.
+
+## API Examples
+
+**Requesting SEV in the VMI spec**
+
+The bellow yaml snippets provide examples of how to request SEV feature in the VMI spec.
+
+*SEV VM with default policy (no attestation):*
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev: {}
+```
+
+*SEV VM with user-defined policy (no attestation):*
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev:
+        policy:
+          - EncryptedState
+```
+
+or alternatively
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev:
+        policy:
+          encryptedState: true
```

Comment:
Do we expect to have multiple policies?

Reply:
Yes, there are multiple policy flags. They are defined in the [SEV spec](https://www.amd.com/system/files/TechDocs/55766_SEV-KM_API_Specification.pdf) (Chapter 3 Guest Policy). You can also find those in the [SEV implementation PR](https://github.com/kubevirt/kubevirt/pull/6166):

```go
type SEVPolicy struct {
	// Sharing keys with other guests is disallowed.
	// Defaults to false.
	// +optional
	NoKeysSharing *bool `json:"noKeysSharing,omitempty"`
	// SEV-ES is required.
	// Defaults to false.
	// +optional
	EncryptedState *bool `json:"encryptedState,omitempty"`
	// Sending the guest to another platform is disallowed.
	// Defaults to false.
	// +optional
	NoSend *bool `json:"noSend,omitempty"`
	// The guest must not be transmitted to another platform that is not in the domain.
	// Defaults to false.
	// +optional
	Domain *bool `json:"domain,omitempty"`
	// The guest must not be transmitted to another platform that is not SEV capable.
	// Defaults to false.
	// +optional
	SEV *bool `json:"sev,omitempty"`
}
```